### PR TITLE
Expand accounts api to expose publicly viewable information for other users

### LINKED ISSUE #5261: Add UserProfile API public view methods
Supports user profile pages on addons-frontend.

First step - API(s) to allow publicly viewable information to be shown ([e.g.](https://addons.mozilla.org/en-US/firefox/user/eviljeff/))

PRD [spec](https://docs.google.com/document/d/1cJ1k8HsAb6taauvJMDLOiYMa1McfUXNgEtax3hoToPc/edit?pli=1#heading=h.t8mr6z8hqji)

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #5261

### CODE REVIEW

File: docs/topics/api/accounts.rst [hunk 1]
```diff
@@ -2,64 +2,83 @@
...
 
-    **Response:**
+    .. _profile-object:
 
-    .. sourcecode:: json
+    :>json int id: The numeric user id.
+    :>json string username: username chosen by the user, used in the profile url. If not set will be a randomly generated string.
+    :>json string name: The name chosen by the user, or the username if not set.
+    :>json array addons_listed: Array holding the :ref:`addons <addon-detail-object>` the user has developed and are listed on this website.
+    :>json float averagerating: The average rating for the `addons_listed`.
```
Comment (diox) [inline docs/topics/api/accounts.rst]: I vote for changing this field name in the API to at least `average_rating` - or something more self-descriptive.

File: docs/topics/api/accounts.rst [hunk 2]
```diff
@@ -2,64 +2,83 @@
...
+    :>json string|null biography: More details about the user.
+    :>json string|null homepage: The user's website.
+    :>json string|null location: The location of the user.
+    :>json string|null occupation: The occupation of the user.
+    :>json string picture_url: URL to a photo of the user, or `/static/img/anon_user.png` if not set.
+    :>json string|null picture_type: the image type (only 'image/png' is supported) if a user defined photo has been provided, or none if no photo has been provided.
+    :>json array reviews: Array holding the :ref:`reviews <review-detail-object>` the user has written.
```
Comment (diox) [inline docs/topics/api/accounts.rst]: This array can be enormous, it should be paginated... Why do we need this anyway? We already have a reviews API, and it allows you to filter by user.

File: docs/topics/api/accounts.rst [hunk 3]
```diff
@@ -2,64 +2,83 @@
...
+If you authenticate and access your own profile (either by specifing your own user_id, or omiting it) the following additional fields are returned.
+If you have `Users:Edit` permission you will see these extra fields for all user profiles.
+
+.. http:get:: /api/v3/accounts/profile/
+
+    .. _self-profile-object:
+
+    :>json string email: Email address used by the user to login and create this account.
+    :>json string|null display_name: The name chosen by the user.
+    :>json array collections: Array holding the collections the user has created.
```
Comment (diox) [inline docs/topics/api/accounts.rst]: Again, this should be paginated, and therefore should be a separate API allowing you to filter by user.

File: docs/topics/api/accounts.rst [hunk 4]
```diff
@@ -2,64 +2,83 @@
...
+------------
+Self Profile
+------------
+
+.. _`self-profile`:
+
+If you authenticate and access your own profile (either by specifing your own user_id, or omiting it) the following additional fields are returned.
+If you have `Users:Edit` permission you will see these extra fields for all user profiles.
+
+.. http:get:: /api/v3/accounts/profile/
```
Comment (diox) [inline docs/topics/api/accounts.rst]: Since we're making backwards-incompatible changes anyway, I'd prefer to follow the usual pattern of doing `/accounts/account/` and `/accounts/account/xxx/` and use `ViewSets`, preparing for a future API where we allow writes and not just reads.

We can redirect the old URL if we want to maintain compatibility.

File: docs/topics/api/accounts.rst (hunk 2 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: It was in the [spec](https://docs.google.com/document/d/1cJ1k8HsAb6taauvJMDLOiYMa1McfUXNgEtax3hoToPc/edit?pli=1#heading=h.t8mr6z8hqji) as something required on the page and I assumed a single API request was desirable.  If it's not needed I can drop (or paginate if it is)

File: docs/topics/api/accounts.rst (hunk 3 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: Collection pages were out of scope for this API (they're on the PRD but all Q3) so I didn't want to delve to much into it.  If pagination isn't enough I'm not sure of the solution here.

File: docs/topics/api/accounts.rst (hunk 2 above)
Comment (diox) [inline docs/topics/api/accounts.rst]: I think it's best to drop it, make a separate API request for this.

File: docs/topics/api/accounts.rst (hunk 3 above)
Comment (diox) [inline docs/topics/api/accounts.rst]: Again, a separate request is preferable here IMHO since there might be a ton of collections returned. Just build the minimal API required to do this. (with filtering like the reviews API, but only implement filtering on `user`, and raise an error if the filter parameter is not present)

Comment (diox) [inline docs/topics/api/accounts.rst]: Could be a separate PR btw

File: docs/topics/api/accounts.rst [hunk 5]
```diff
@@ -2,64 +2,83 @@
...
+.. http:get:: /api/v3/accounts/profile/(int:user_id)/
 
-    **Response:**
+    .. _profile-object:
 
-    .. sourcecode:: json
+    :>json int id: The numeric user id.
+    :>json string username: username chosen by the user, used in the profile url. If not set will be a randomly generated string.
+    :>json string name: The name chosen by the user, or the username if not set.
+    :>json array addons_listed: Array holding the :ref:`addons <addon-detail-object>` the user has developed and are listed on this website.
```
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: Btw, this array could be similarly enormous - and there is a very functional addons API

**Comment by eviljeff** [comment]
I think there's a decision needed on whether we want a profile page to involve potentially 4 separate API requests (account, addons_listed, reviews, collections) or some should be included in the main response.

File: docs/topics/api/accounts.rst (hunk 5 above)
Comment (diox) [inline docs/topics/api/accounts.rst]: Yep. It's bad too IMHO.

**Comment by diox** [comment]
Duplicating info into several APIs to avoid multiple calls is fine... up to a point. The point, IMHO, is where you start having to include lists of things without any limits. It's why the addons API detail endpoint includes the current version (and its files) information but not all versions, for instance.

**Comment by eviljeff** [comment]
Removing the fields is easiest but they could all be limited to, say, the 5 most addons|reviews|collections.
@kumar303 do you have a preference from the client-side perspective?

File: docs/topics/api/accounts.rst (hunk 2 above)
Comment (kumar303) [inline docs/topics/api/accounts.rst]: We don't need this in the profile response. As mentioned, we can make a request to the reviews endpoint to get all reviews for the user.

**Comment by kumar303** [comment]
I would prefer to follow what @diox is saying and keep the APIs separate from eachother -- i.e. require the client to make multiple requests. I think it would be easier to combine API results into a single endpoint *later* when we know for sure that we need to solve a specific performance problem in the frontend. For starters, I think it's fine to make separate requests.

File: docs/topics/api/accounts.rst (hunk 1 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: https://github.com/mozilla/addons-server/pull/5318/commits/4e2e7f15a9a7803401e7f795509fb3fc6a991a72

File: docs/topics/api/accounts.rst (hunk 2 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: https://github.com/mozilla/addons-server/pull/5318/commits/4e2e7f15a9a7803401e7f795509fb3fc6a991a72
I included `num_addons_listed` from `users.num_addons_listed` though so we can replicate the current profile page where average rating and count is display in the table and the add-ons themselves below.

File: docs/topics/api/accounts.rst (hunk 3 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: removed in https://github.com/mozilla/addons-server/pull/5318/commits/4e2e7f15a9a7803401e7f795509fb3fc6a991a72
I'll follow-up to re-add the minimal collections API

File: docs/topics/api/accounts.rst (hunk 4 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: Fixed in the other commits.

File: docs/topics/api/accounts.rst [hunk 6]
```diff
@@ -2,64 +2,77 @@
...
+------------
+Self Account
+------------
+
+.. _`self-account`:
+
+If you authenticate and access your own account (either by specifing your own user_id, or omiting it) the following additional fields are returned.
+If you have `Users:Edit` permission you will see these extra fields for all user accounts.
+
+.. http:get:: /api/v3/accounts/account/
```
Comment (diox) [inline docs/topics/api/accounts.rst]: As said on IRC, I really don't think that's a good idea. `/accounts/account/` should be reserved for listing accounts. That's how DRF `SimpleRouter` does things by default, and we already rely on that behavior for the reviews API. Making an exception here makes things inconsistent and will add complexity later if we ever need to implement accounts listing.

Comment (diox) [inline docs/topics/api/accounts.rst]: Just having `/accounts/profile/` redirect to `/accounts/account/<userid>/` should be enough - or, if we really want a specific endpoint, make `/accounts/profile/` a simple view that calls AccountViewSet detail method behind the scenes instead of a redirect.

Comment (eviljeff) [inline docs/topics/api/accounts.rst]: https://github.com/mozilla/addons-server/pull/5318/commits/0a11b0b0043458e96eea07eea1a0a4be6817504c should do it.

File: src/olympia/accounts/views.py [hunk 7]
```diff
@@ -366,13 +370,32 @@ def delete(self, request, *args, **kwargs):
...
+        if (self.self_view or
+                acl.action_allowed_user(self.request.user,
+                                        amo.permissions.USERS_EDIT)):
+            return UserProfileSerializer
+        else:
+            return PublicUserProfileSerializer
+
+    def get_object(self):
+        lookup_value = self.kwargs.get(self.lookup_url_kwarg)
+        if self.request.user.is_authenticated():
```
Comment (diox) [inline src/olympia/accounts/views.py]: Proposed patch to make this more consistent with the rest of the APIs, fixing the `authentication_classes` as well: 
https://gist.github.com/diox/690f8391bf85db502fdfd3a0eff5110a

This makes the profile url a simple method on the viewset, to stop having to manually instantiate the viewset in `urls.py`

Comment (eviljeff) [inline src/olympia/accounts/views.py]: The patch drops the regex on the lookup but that's fairly harmless I suppose
